README.md
# CraftyBox2D
It's a component that facilitates the use of the Physics Engine Box2D in [Crafty](http://craftyjs.com/). The Box2D implementation 
that the component use is [box2dweb](http://code.google.com/p/box2dweb/).

***

##Using CraftyBox2D
First you must init the box2dWorld as follow
<code>
	Crafty.box2D.init(0, 10, 32, true);
</code>

The above code initialize the world with 0 for the world gravity in the x-axis, 10 for the world gravity in the y-axis
32 for the pixel-to-meter ratio, and allow the world sleep. The init method also attach the world.steep() function tho
the stage "EnterFrame" event

Once the world has been initialized, you can start to add entitys with the Box2D component
<code>
	 var floor = Crafty.e("2D, Canvas, Box2D")
		.attr({ x: 0, y: 0})
		.box2d({
			bodyType: 'static',
			shape: [
					[0, 300],
					[400, 300]
				]
		});
						
    var box = Crafty.e("2D, Canvas, Color, Box2D")
		 .attr({ x: 0, y: 0, w:20, h:20})
		 .color("#FF0000")
		 .box2d({
				bodyType: 'dynamic'
		 });
</code>

Above first we create the floor, set the bodyType to static, and set the collition box in the botom of the stage
then we create a box with 20x20 pixels, red color and set the body type to dynamic and left the collition area to
the default values. Running the game a see the box falling to the floor.

the are more options that you can set, please see the demo game code for more.

changelog.txt
v0.1 - 28/01/2012

* First release

test.js
window.onload = function () {
    gameInit();
};

gameInit = function(){
	//start crafty
    Crafty.init(400, 300);
    Crafty.canvas.init();
	Crafty.box2D.init(0, 10, 32, true);
	 
    var floor = Crafty.e("2D, Canvas, Box2D")
				.attr({ x: 0, y: 0})
				.box2d({
					bodyType: 'static',
					shape: [
								[0, 300],
								[400, 300]
							]
					});
						
    var box = Crafty.e("2D, Canvas, Color, Box2D")
		 .attr({ x: 0, y: 0, w:20, h:20})
		 .color("#FF0000")
		 .box2d({
				bodyType: 'dynamic'
		 });
}

box2d.js
var   b2Vec2 = Box2D.Common.Math.b2Vec2
         	,	b2BodyDef = Box2D.Dynamics.b2BodyDef
         	,	b2Body = Box2D.Dynamics.b2Body
         	,	b2FixtureDef = Box2D.Dynamics.b2FixtureDef
         	,	b2Fixture = Box2D.Dynamics.b2Fixture
         	,	b2World = Box2D.Dynamics.b2World
         	,	b2MassData = Box2D.Collision.Shapes.b2MassData
         	,	b2PolygonShape = Box2D.Collision.Shapes.b2PolygonShape
         	,	b2CircleShape = Box2D.Collision.Shapes.b2CircleShape
         	,	b2DebugDraw = Box2D.Dynamics.b2DebugDraw
			,	b2MouseJointDef = Box2D.Dynamics.Joints.b2MouseJointDef
			,	b2PrismaticJointDef = Box2D.Dynamics.Joints.b2PrismaticJointDef
            ;
			
/**@
* #Box2D
* @category Box2D
* Component that adds the Box2D physics engine capabilities to the entities.
*/
Crafty.c("Box2D", {	
	/**@
	* #.body
	* The 'b2Body' element of the entity.	
	*/
	body : null,
	
	/**@
	* #.fixtures
	* Array than contains the b2Fixtures of the entity.	
	*/
	fixtures : null,
	
	init: function () {
		this.requires("2D");
		
		if (!Crafty.box2D.world) {
			Crafty.box2D.init(0, 10, 55, true);
		}
	},
	/**@
	* #.box2d
	* @comp Box2D		
	* @sign public void .box2d(Object obj)
	* @param obj - Object with the bodyType(dynamic, static, kinematic) and fixture data to make
	* the fist fixture of the body
	*
	* Create the b2Body and link to the crafty entity, it only need the bodyType
	* ie: .box2d({bodyType: 'solid'}) 
	* this will create a b2Body and make a fixture with the default, the other params of the object,
	* are the same of the method .addFixture
	* 
	*/
	box2d: function(obj){
		var world = Crafty.box2D.world;
		var PTM_RATIO =	Crafty.box2D.PTM_RATIO;
		var fixDef;
		var vertexCount = 0;
		this.fixtures = [];
		
		var BodyDef = new b2BodyDef;
		
		if(obj.bodyType === "dynamic"){
			BodyDef.type = b2Body.b2_dynamicBody;
		}else if (obj.bodyType === "static"){
			BodyDef.type = b2Body.b2_staticBody;
		}else{
			BodyDef.type = b2Body.b2_kinematicBody;
		}
		
		
		BodyDef.position.Set(this._x/PTM_RATIO, this._y/PTM_RATIO);
		BodyDef.userData = this;
		
		this.body = world.CreateBody(BodyDef);
		
		this.addFixture(obj);
		
		return this;
	},
	/**@
	* #.addFixture
	* @comp Box2D		
	* @sign public void .addFixture(Object fixture)
	* @param fixture - Object with the B2FixtureDef or the propierties to make 
	*	a B2Fixture (Density, Friction, Restitution, Shape, etc)
	*
	* Add a fixture to te body of the entity, you can pass a complete B2FixtureDef for more control
	* ie: .addFixture({fixDef: B2FixtureDef}) or pass the propierties what you want 
    * ie: .addFixture(
	*				  {
	*					density: Number, 
	*					friction: Number, 
	*					restitution: Number, 
	*					shape: Array with the shape or String with the shape you want('circle', 'box')
    *				  }) 	
	*
	* If you pass a empty obj, the metod will create a fixture with default values, when the 
	* fixture is made, it's attached to the body and add to the fixtures Array
	*/
	addFixture : function(fixtureDef){
		var PTM_RATIO =	Crafty.box2D.PTM_RATIO;
		
		if(fixtureDef.fixDef){
			fixDef = obj.fixDef;
		}else{
			fixDef = new b2FixtureDef;
			fixDef.density = (!isNaN(fixtureDef.density)) ? fixtureDef.density : 1;
			fixDef.friction = (!isNaN(fixtureDef.friction)) ? fixtureDef.friction : 0.5;
			fixDef.restitution = (!isNaN(fixtureDef.restitution)) ? fixtureDef.restitution : 0.2;
			fixDef.shape = new b2PolygonShape;
			
			//console.log(fixDef);
			if(!fixtureDef.shape || typeof fixtureDef.shape === "string"){
				if(fixtureDef.shape === "circle"){
					fixDef.shape = new b2CircleShape(
											 this._w/PTM_RATIO/2
										   );
					
					fixDef.shape.SetLocalPosition(
								new b2Vec2(
										this._w/PTM_RATIO/2, 
										this._h/PTM_RATIO/2
									  )
							);
				
				}else{
					vertexCount = 2;
					fixDef.shape.SetAsOrientedBox(
											this._w/PTM_RATIO/2, 
											this._h/PTM_RATIO/2, 
											new b2Vec2(
														this._w/PTM_RATIO/2, 
														this._h/PTM_RATIO/2
													  )
											);
				}
			}else{
				vertexCount = fixtureDef.shape.length;
				var shapeArray = [];
				for(var i = 0; i < vertexCount; i++){
					var vector = fixtureDef.shape[i]
					shapeArray.push(new b2Vec2(vector[0]/PTM_RATIO, vector[1]/PTM_RATIO));
				}
				fixDef.shape.SetAsArray(
									shapeArray, vertexCount
								);
			}
			
			this.fixtures.push(this.body.CreateFixture(fixDef));
			
			return this;
		}
	}
});

/**@
* #Crafty.box2D
* @category Box2D
* Collection of methods to init Box2D World.
*/
Crafty.extend({
	box2D: {
		ShowBox2DDebug : false,
	/**@
		* #Crafty.box2D.world
		* @comp Crafty.box2D
		* This will return the b2World element.
		*/
		world : null,
		
	/**@
		* #Crafty.box2D.PTM_RATIO
		* @comp Crafty.box2D
		* This will return the pixel-to-meter ratio used to draw the b2World.
		*/
		PTM_RATIO : null,
		
		/**@
		* #Crafty.box2D.init
		* @comp Crafty.box2D		
		* @sign public void Crafty.box2D.init(Number gx, Number gy, Number ptm_ratio, Boolean doSleep)
		* @param gx - gravity force of the x-axis
		* @param gy - gravity force of the y-axis
		* @param ptm_ratio - pixel-to-meter ratio
		* @param doSleep permit the Box2D world sleep
		* Creates a b2World element and bind the Box2D 'step' to Crafty EnterFrame funct
		* Must be called before any entities with the Box2D component can be drawn.
		*
		* This method will automatically be called if no `Crafty.canvas.b2World` is
		* found.
		*/
		init: function (gx, gy, ptm_ratio, doSleep) {
			var _world = new b2World(
							   new b2Vec2(gx, gy)    //gravity
							,  doSleep                 //allow sleep
						 );
		 
			var _PTM_RATIO = ptm_ratio;
			
			Crafty.bind("EnterFrame", function() {
				_world.Step(
					   1 / 30   //frame-rate
					,  8       //velocity iterations
					,  3       //position iterations
				 );
				 
				for(var b = _world.GetBodyList(); b; b=b.GetNext()) {    
					if (b.GetUserData()) {
						var sprite = b.GetUserData(); 
						sprite.attr(
									{
										x: b.GetPosition().x * _PTM_RATIO, 
										y:b.GetPosition().y * _PTM_RATIO
									}					
							);
						sprite.rotation = Crafty.math.radToDeg(b.GetAngle());
									
					}        
				}
				
				if(Crafty.box2D.ShowBox2DDebug){
					_world.DrawDebugData();
				}
				_world.ClearForces();
				
			});
			
			Crafty.box2D.world = _world;
			Crafty.box2D.PTM_RATIO = _PTM_RATIO;
		},
		
		showDebugInfo : function(){
			var _world = Crafty.box2D.world;
			var _PTM_RATIO = Crafty.box2D.PTM_RATIO;
			
			if (Crafty.support.canvas) {				
				var c = document.createElement("canvas");
				c.id = "Box2DCanvasDebug";
				c.width = Crafty.viewport.width;
				c.height = Crafty.viewport.height;
				c.style.position = 'absolute';
				c.style.left = "0px";
				c.style.top = "0px";

				Crafty.stage.elem.appendChild(c);
				
				var debugDraw = new b2DebugDraw();
				debugDraw.SetSprite(c.getContext('2d'));
				debugDraw.SetDrawScale(_PTM_RATIO);
				debugDraw.SetFillAlpha(0.7);
				debugDraw.SetLineThickness(1.0);
				debugDraw.SetFlags(b2DebugDraw.e_shapeBit | b2DebugDraw.e_jointBit);
				_world.SetDebugDraw(debugDraw);				
				Crafty.box2D.ShowBox2DDebug = true;
			}else{
				Crafty.box2D.ShowBox2DDebug = false;
			}
		}
	}
});

demo2.js
window.onload = function () {
    gameInit();
};

gameInit = function(){
	//start crafty
        var stageW = 500;
        var stageH = 400;
        Crafty.init(stageW, stageH);
        Crafty.canvas.init();
        var gx=0;
        var gy=10;
        var ptm_ratio=32;
        var doSleep=true;
        Crafty.box2D.init(gx, gy, ptm_ratio, doSleep);
        Crafty.background("#eee");
        
        var power_bar_count=16;

        Crafty.c("PowerBar", {

                init: function() {
                        this.attr({ x: (stageW-b_size*power_bar_count)/2, y: 20, z: 1, w:1, h:b_size })
                        .color("#fff") 
                },
                

                PowerBar: function(speed) {
                        this.attr({ x: (stageW-b_size*power_bar_count)/2, y: 20, z: 1, w:speed, h:b_size })
                        return this;
                }
        });
             
        var floor = Crafty.e("2D, Canvas, Box2D")
                                    .attr({ x: 0, y: 0})
                                    .box2d({
                                            bodyType: 'static',
                                            shape: [
                                                                    [0, stageH],
                                                                    [stageW, stageH]
                                                            ]
                                            });
        var blck_cnt = 20;
        var b_size = 10;
        var speed = 1;
        var speedMax = b_size*power_bar_count;
        var speedUp = false;
        
        
        for (var x =0; x<blck_cnt/2; x++){
                for (var y =0; y<blck_cnt/2; y++){ 
                var box = Crafty.e("2D, Canvas, Color, Box2D")
                                .attr({ x: (stageW-(x*b_size))/2, y: stageH-(y*b_size)-10, w:b_size, h:b_size})
                                .color("#FF0000")
                                .box2d({
                                                   bodyType: 'dynamic',
                                                   density : 10,
                                                   friction : 30,
                                                   restitution : 0.1
                                });
                }
        }
        
        var PBar = Crafty.e("2D, Canvas, Color, PowerBar")   
        
        var dropper = Crafty.e("2D, Canvas, Color, Mouse")
                .attr({ x: (stageW-(b_size*power_bar_count))/2, y: 0, z: 3, w:b_size*power_bar_count, h:b_size*2 })
                .color("#ccc")
                .bind("EnterFrame", function(){
                        if(speedUp===true){
                                if(speed<speedMax){
                                        speed++;
                                }
                                else{
                                        speed =1;
                                }
                        PBar.PowerBar(speed);
                        }
                })
                .bind("MouseDown", function() {
                        console.log("start at"+speed);
                        speedUp = true;        
                })
                .bind("MouseUp", function() {
                        console.log("release at"+speed);
                        var size = Math.ceil((speed/speedMax)*4);
                        var drop = Crafty.e("2D, Canvas, Color, Box2D")
                                        .attr({ x: (stageW-b_size*2)/2, y: 0, z: 1, w:size*b_size, h:size*b_size })
                                        .color("#0ff")
                                        .box2d({
                                                        bodyType: 'dynamic',
                                                        density : speed,
                                                        friction : 30,
                                                        restitution : 0.5
                                                });
                        speedUp = false;
                })
                ;
                
                
        var txt = Crafty.e("2D, Canvas, Text")
                .attr({ x: (stageW-(b_size*power_bar_count)+15)/2, y: -5, z: 3, w:b_size*2, h:b_size*2 })
                .text("Click Here")
                
}


demo3.js
/* Author: @shogoki_vnz
*/
var world,
	PTM_RATIO = 32,
	w = 600,
	h = 400,
	hw = w/2,
	hh = h/2,
	mouseJoint,	
	ctx = this,
	isMouseDown = false,
	mouseX,
	mouseY,
	mousePVec,
	selectedBody;
	
window.onload = function () {
    gameInit();
};

gameInit = function(){
	//start crafty
    Crafty.init(w, h);
    Crafty.canvas.init();
	//Init the box2d world, gx = 0, gy = 10
	Crafty.box2D.init(0, 10, PTM_RATIO, true);
	world = Crafty.box2D.world;
	//Start the Box2D debugger
	Crafty.box2D.showDebugInfo();  
	
	Crafty.scene("main", function () {				
        generateWorld();     
    });
	
	Crafty.scene("main");
}

generateWorld = function() {	
					
   var walls = Crafty.e("2D, Canvas, Box2D")
				.attr({ x: 0, y: 0})
				.box2d({
						bodyType: 'static',
						density : 1.0,
						friction : 10,
						restitution : 0,
						shape: [
									[0, 0],
									[w, 0],
									[w, 10],
									[0, 10]
								]
						});	
	
	walls.addFixture(
						{
							bodyType: 'static',
							density : 1.0,
							friction : 10,
							restitution : 0,
							shape: [
										[0, 0],
										[10, 0],
										[10, h],
										[0, h]
									]
							
						}
					);
					
	walls.addFixture(
						{
							bodyType: 'static',
							density : 1.0,
							friction : 10,
							restitution : 0,
							shape: [
										[(w-10), 0],
										[w, 0],
										[w, h],
										[(w-10), h],
									]
							
						}
					);
					
	walls.addFixture(
						{
							bodyType: 'static',
							density : 1.0,
							friction : 10,
							restitution : 0,
							shape: [
										[0, (h-10)],
										[w, (h-10)],
										[w, h],
										[0, h]
									]
							
						}
					);
	
	
	genterateChars();	
}

genterateChars = function(){	 
	 for(var i = 0; i < 10; ++i) {
		var _w, _h, shape;
		
		if(Math.random() > 0.5) {
			_w = Crafty.math.randomInt(PTM_RATIO, PTM_RATIO*2);
			_h = Crafty.math.randomInt(PTM_RATIO, PTM_RATIO*2);
			shape = "box";
		}else{
			_w = Crafty.math.randomInt(PTM_RATIO, PTM_RATIO*2);
			_h = _w;
			shape = "circle";
		}
		
		var fallingElement =  Crafty.e("2D, Canvas, Color, Mouse, Box2D")
							.origin("center")
							.color("#fff")
							.attr({ 
								x: Crafty.math.randomInt(30, w-100), 
								y: 0, 
								h: _w, 
								w: _h
							})
							.box2d({
									bodyType: 'dynamic',
									density : 1.0,
									friction : 2,
									restitution : 0.2,
									shape: shape
							});	
	 }
	 
	Crafty.addEvent(ctx, "mousedown", 
									function(e) {
										isMouseDown = true;
										mousemoved(e);
										Crafty.addEvent(ctx, "mousemove", mousemoved)
									 });
         
    Crafty.addEvent(ctx, "mouseup", function(){
									Crafty.removeEvent(ctx, "mousemove", mousemoved);
									isMouseDown = false;
									mouseX = undefined;
									mouseY = undefined;
								});
								
	Crafty.bind("EnterFrame", onEnterFrame);
}

mousemoved = function(e) {
	mouseX = e.clientX / PTM_RATIO;
	mouseY = e.clientY / PTM_RATIO;
};

getBodyAtMouse = function() {
	mousePVec = new b2Vec2(mouseX, mouseY);
	var aabb = new Box2D.Collision.b2AABB();
	aabb.lowerBound.Set(mouseX - 0.001, mouseY - 0.001);
	aabb.upperBound.Set(mouseX + 0.001, mouseY + 0.001);
	
	// Query the world for overlapping shapes.

	selectedBody = null;
	world.QueryAABB(getBodyCB, aabb);
	return selectedBody;
 }

getBodyCB = function(fixture) {
	if(fixture.GetBody().GetType() !== b2Body.b2_staticBody) {
	   if(fixture.GetShape().TestPoint(fixture.GetBody().GetTransform(), mousePVec)) {
		  selectedBody = fixture.GetBody();
		  return false;
	   }
	}
	return true;
 }

onEnterFrame = function() {
	if(isMouseDown && (!mouseJoint)) {
               var body = getBodyAtMouse();
               if(body) {
                  var md = new b2MouseJointDef();
                  md.bodyA = world.GetGroundBody();
                  md.bodyB = body;
                  md.target.Set(mouseX, mouseY);
                  md.collideConnected = true;
                  md.maxForce = 300.0 * body.GetMass();
                  mouseJoint = world.CreateJoint(md);
                  body.SetAwake(true);
               }
            }
            
	if(mouseJoint) {
	   if(isMouseDown) {
		  mouseJoint.SetTarget(new b2Vec2(mouseX, mouseY));
	   } else {
		  world.DestroyJoint(mouseJoint);
		  mouseJoint = null;
	   }
	}
}


pong.js
/* Author: @shogoki_vnz
	this game is based in http://www.raywenderlich.com/475/how-to-create-a-simple-breakout-game-with-box2d-and-cocos2d-tutorial-part-12	
*/
var world,
	PTM_RATIO = 32,
	w = 800,
	h = 600,
	hw = w/2,
	hh = h/2,
	walls,
	ball,
	paddle,
	jointDef,
	_contacts = [],
	toDestroy = [],
	totalBlocks,	
	ctx = this;
	
window.onload = function () {
    gameInit();
};

gameInit = function(){
	//start crafty
    Crafty.init(w, h);
    Crafty.canvas.init();
	Crafty.box2D.init(0, 0, PTM_RATIO, true);
	world = Crafty.box2D.world;
	//Uncoment this tho see the Box2D DebugInfo
	//Crafty.box2D.showDebugInfo();
	
	//Add the contactlistener and bind the Crafty EnterFrame	
	var contactListener = new Box2D.Dynamics.b2ContactListener;
    contactListener.BeginContact = function(contact)
								   {
										var myContact = { 
															fixtureA: contact.GetFixtureA(), 
															fixtureB: contact.GetFixtureB() 
														};
										_contacts.push(myContact);
								   };
								   
	contactListener.EndContact = function(contact)
								   {										
										var myContact = { 
															fixtureA: contact.GetFixtureA(), 
															fixtureB: contact.GetFixtureB() 
														};
										
										var totalContacts = _contacts.length;
										for(var i = 0; i < totalContacts; ++i){
											if ((_contacts[i].fixtureA == myContact.fixtureA) && (_contacts[i].fixtureB == myContact.fixtureB)) {
												_contacts.splice(i, 1);
												return;
											}
										}
								   };
								   
	world.SetContactListener(contactListener);	
	
	
	Crafty.sprite(15, "img/ball.png", {
		ball: [0, 0]
	});
	
	
    Crafty.scene("loading", function () {        
        Crafty.load(["img/ball.png"], function () {	
			createPlayerComponents();             
        });       
        Crafty.background("#000");
        Crafty.e("2D, DOM, Text").attr({ w: 100, h: 20, x: hw-50, y: hh-10 })
                .text("Loading")
                .css({ "text-align": "center", "color": "#ffffff"});
    });   
    Crafty.scene("loading");
	
	Crafty.scene("main", function () {				
        generateWorld();     
    });
	
	Crafty.scene("gameover", function () {		
        Crafty.e("2D, DOM, Text").attr({ w: 180, h: 20, x: hw-80, y: hh-10 })
                .text("Sorry Game Over Bro...<br/>Click on the stage to restart")
                .css({ "text-align": "center", "color": "#ffffff"});
		destroyWorld();
		Crafty.addEvent(ctx, "click", restartgame)
    }); 

	Crafty.scene("gamecomplete", function () {		
        Crafty.e("2D, DOM, Text").attr({ w: 180, h: 20, x: hw-80, y: hh-10 })
                .text("Congratulations You Win!<br/>Click on the stage to restart")
                .css({ "text-align": "center", "color": "#ffffff"});
		destroyWorld();
		Crafty.addEvent(ctx, "click", restartgame)
    }); 
}



createPlayerComponents = function(){
	Crafty.c("PaddleControls", {
		init: function() {
		  this.requires("Keyboard");		  
		  return this.speed = 3;
		},
		paddleControls: function(speed) {			
		  if (speed) {
			this.speed = speed;
		  }
		  this.bind("EnterFrame", function() {			
			if (this.disableControls) {
			  return;
			}
			var dx = 0;
			
			if (this.isDown("D")) {
			  dx = this.speed;
			}
			if (this.isDown("A")) {
			  dx = -1 * this.speed;
			}
			
			if (dx != null) {
			  return this.body.ApplyImpulse(new b2Vec2(dx/PTM_RATIO, 0), this.body.GetWorldCenter());
			}
		  });
		  return this;
		}
	  });
	  
	Crafty.c("Colorme", {
		colors : null,
		init: function() {
		  this.requires("Color");
		  this.colors = ["#fff", "#00FF00", "#0000FF" ,"#FF0000"];
		},
		colorme: function(index) {	
		  this.color(this.colors[index]);
		  return this;
		}
	  });

	Crafty.c("GameBlock", {
		isBlock : true,
		id : '',
		remain : 0,
		gameBlock: function(_id, _remain) {	
		  this.id = _id;
		  this.remain = _remain;
		  return this;
		}
	 });
	  
	Crafty.scene("main");
}

generateWorld = function() {	
					
    walls = Crafty.e("2D, Canvas, Color, Box2D")
				.attr({ x: 0, y: 0})
				.box2d({
						bodyType: 'static',
						density : 1.0,
						friction : 0,
						restitution : 0,
						shape: [
									[0, 0],
									[w, 0]
								]
						});	
	
	walls.addFixture(
						{
							bodyType: 'static',
							density : 1.0,
							friction : 0,
							restitution : 0,
							shape: [
										[0, 0],
										[0, h]
									]
							
						}
					);
					
	walls.addFixture(
						{
							bodyType: 'static',
							density : 1.0,
							friction : 0,
							restitution : 0,
							shape: [
										[w, 0],
										[w, h]
									]
							
						}
					);
					
	walls.addFixture(
						{
							bodyType: 'static',
							density : 1.0,
							friction : 0,
							restitution : 0,
							shape: [
										[0, h],
										[w, h]
									]
							
						}
					);
	
	
	genterateChars();	
}

genterateChars = function(){
	//create our ball
	 ball = Crafty.e("2D, Canvas, ball, Box2D")
			.attr({ x: hw-15, y: 575, z: 1, isBall:true})
			.box2d({
					bodyType: 'dynamic',
					density : 1,
					friction : 0.2,
					restitution : 1,
					shape : 'circle'
				});	
					
	ball.body.SetBullet(true);
	var	fx = (Math.random() > 0.5) ? 2 : -2;
	var force = new b2Vec2(fx, 1.5);
	ball.body.ApplyImpulse(force, ball.body.GetWorldCenter());
	
	paddle = Crafty.e("2D, Canvas, Color, Box2D, PaddleControls")
		  .attr({ x: hw-35, y: 580, z: 1, w:270, h:5 })
		  .color("#fff")		  
		  .box2d({
						bodyType: 'dynamic',
						density : 8,
						friction : 1,
						restitution : 0
					})
		  .paddleControls(70)
		  .bind("EnterFrame", 
						function() {
							var maxSpeed = 10;							
 
							var velocity = this.body.GetLinearVelocity();
							var speed = velocity.Length();							
							if (speed > maxSpeed) {
								this.body.SetLinearDamping(0.5);
							} else if (speed < maxSpeed) {
								this.body.SetLinearDamping(0.0);
							}
						}			
					);
	
	// Restrict paddle along the x axis
	jointDef = new b2PrismaticJointDef;
	var worldAxis = new b2Vec2(1, 0);
	jointDef.collideConnected = true;
	jointDef.Initialize(paddle.body, walls.body, 
	  paddle.body.GetWorldCenter(), worldAxis);
	world.CreateJoint(jointDef);
	
	//Adding some blocks to play
	var blocksByColumns = 13;
	totalBlocks = 0;
	for(var i = 0; i < 8; i++) {
		var padding=5;
		var blockw = 50;
		blocksByColumns--;
		for(var j = 0; j < blocksByColumns; ++j){	
			totalBlocks++
			var xOffset = (hw-(blocksByColumns*((blockw+padding)/2)))+((blockw+padding)*j);
			var rand = Math.random();
			var maxTouchs = Crafty.math.randomInt(1, 4);
			var posy = 70+(15*i);
			var block = Crafty.e("2D, Canvas, Colorme, Box2D, GameBlock")
			  .attr({ x: xOffset, y: posy, w:blockw, h:5})
			  .colorme(maxTouchs-1)
			  .gameBlock("block"+j+'_'+i, maxTouchs)
			  .box2d({
							bodyType: 'solid',
							density : 10,
							friction : 0,
							restitution : 1
						});
		}	 
	}	
	
	Crafty.bind("EnterFrame", onEnterFrame);
}

addToDestroy = function(otherBody){
	
	var totalDestroy = toDestroy.length;
	if(totalDestroy > 0){
		for(var i = 0; i < totalDestroy; ++i){	
			var body = toDestroy[i];
			
			if (body.GetUserData() && otherBody.GetUserData()) {
				var sprite = body.GetUserData();
				var otherSprite = otherBody.GetUserData();
				if (sprite.id !== otherSprite.id){
					toDestroy.push(otherBody);
				}			
			}		
		}
	}else{
		toDestroy.push(otherBody);
	}
}

destroyWorld = function (){	
	for(var b = world.GetBodyList(); b; b=b.GetNext()) {    
		/*if (b.GetUserData()) {
			var sprite = b.GetUserData(); 
			world.DestroyBody(sprite.body);
			sprite.destroy();	
			sprite = null;			
		}*/
		world.DestroyBody(b);      
	}
	_contacts = [];
	toDestroy = [];
	Crafty.unbind("EnterFrame", onEnterFrame);
}

restartgame = function(){
	Crafty.removeEvent(ctx, "click", restartgame)
	Crafty.scene("main");
}

onEnterFrame = function() {
	var totalContacts = _contacts.length;
	for(var i = 0; i < totalContacts; ++i){	
		var contact = _contacts[i];
		if ((contact.fixtureA === walls.fixtures[3] && contact.fixtureB === ball.fixtures[0]) ||
			(contact.fixtureA === ball.fixtures[0] && contact.fixtureB === walls.fixtures[3])) {
			Crafty.scene("gameover");
		}
		
		var bodyA = contact.fixtureA.GetBody();
		var bodyB = contact.fixtureB.GetBody();
		if ((bodyA.GetUserData()) && (bodyB.GetUserData())) {
			var spriteA = bodyA.GetUserData();
			var spriteB = bodyB.GetUserData();
	 
			// Sprite A = ball, Sprite B = Block
			if (spriteA.isBall && spriteB.isBlock) {
				bodyB.GetUserData().remain--;
				if(bodyB.GetUserData().remain <= 0){
					addToDestroy(bodyB);
				}else{	
					bodyB.GetUserData().colorme(bodyB.GetUserData().remain-1);
				}
			}
			// Sprite B = block, Sprite A = ball
			else if (spriteA.isBlock && spriteB.isBall) {
				bodyA.GetUserData().remain--;
				if(bodyA.GetUserData().remain <= 0){
					addToDestroy(bodyA);
				}else{						
					bodyA.GetUserData().colorme(bodyA.GetUserData().remain-1);
				}
			}        
		} 
	}
	
	var totalDestroy = toDestroy.length;
	for(var j = 0; j < totalDestroy; ++j) {
		var body = toDestroy[j];     
		if (body.GetUserData()) {
			var sprite = body.GetUserData();
			sprite.destroy();		
			sprite = null;
		}
		world.DestroyBody(body);
		totalBlocks--;
		if(totalBlocks === 0){
			Crafty.scene("gamecomplete");
		}
	}
	
	toDestroy = [];
}
